src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp

```diff
@@ -0,0 +1,140 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ONNXShapeHelper.hpp"
+
+using namespace mlir;
+
+struct ONNXArgMaxOpLowering : public ConversionPattern {
+  ONNXArgMaxOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXArgMaxOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    // Gather info.
+    auto loc = op->getLoc();
+    ONNXArgMaxOpAdaptor operandAdaptor(operands);
+    ONNXArgMaxOp argMaxOp = llvm::cast<ONNXArgMaxOp>(op);
+
+    // shape helper
+    ONNXArgMaxOpShapeHelper shapeHelper(&argMaxOp, &rewriter);
+
+    auto shapecomputed = shapeHelper.Compute(operandAdaptor);
+    (void)shapecomputed;
+    assert(!failed(shapecomputed) && "expected to succeed");
+
+    // reduced output
+    auto reducedMemRefType = convertToMemRefType(*op->result_type_begin());
+    auto reducedElementType = reducedMemRefType.getElementType();
+    auto reducedMemrefShape = reducedMemRefType.getShape();
+    int64_t reducedRank = reducedMemRefType.getRank();
+
+    // data input
+    auto data = operandAdaptor.data();
+    auto dataType = data.getType().cast<MemRefType>();
+    int64_t dataRank = dataType.getRank();
+
+    // axis & keepdims attribute
+    int64_t axis = argMaxOp.axis();
+    assert(axis >= -dataRank && axis <= dataRank - 1);
+    axis = axis >= 0 ? axis : (dataRank + axis);
+
+    int64_t keepdims = argMaxOp.keepdims();
+    bool isKeepdims = (keepdims == 1) ? true : false;
+
+    // Get type information
+    llvm::SmallVector<int64_t, 1> axes;
+    axes.push_back(axis);
+    std::map<int64_t, int64_t> outInDimMap =
+        getReductionMapping(dataType, llvm::makeArrayRef(axes), isKeepdims);
+
+    // Insert alloc and dealloc
+    Value alloc = insertAllocAndDeallocSimple(
+        rewriter, op, reducedMemRefType, loc, shapeHelper.dimsForOutput(0));
+
+    // Constant Value
+    auto minusOne = emitConstantOp(rewriter, loc, reducedElementType, -1);
+    auto zero = emitConstantOp(rewriter, loc, reducedElementType, 0);
+    auto zeroIndex = rewriter.create<ConstantIndexOp>(loc, 0);
+
+    // 1. Krnl loops to initialize the result.
+    BuildKrnlLoop initLoops(rewriter, loc, reducedRank);
+    initLoops.createDefineOp();
+    initLoops.pushAllBounds(shapeHelper.dimsForOutput(0));
+    initLoops.createIterateOp();
+    rewriter.setInsertionPointToStart(initLoops.getIterateBlock());
+
+    // Handle the operation:
+    SmallVector<Value, 4> loopIVs;
+    for (auto arg : initLoops.getAllInductionVar()) {
+      loopIVs.push_back(arg);
+    }
+
+    rewriter.create<KrnlStoreOp>(loc, minusOne, alloc, loopIVs);
+
+    auto parent = rewriter.getInsertionPoint()->getParentOp();
+    rewriter.setInsertionPointAfter(parent);
```

Comment:
Change to use `saveInsertionPoint` & `restoreInsertionPoint`. Thanks!

Reply:
I meant to use saveInsertionPoint() before you change insertion point to the loop body. The intention is to avoid the assumption the insert point is the parent block of current insertion place.